fish-rust/src/builtins/path.rs

```diff
@@ -0,0 +1,990 @@
+            t if t == "group" => Self::GROUP,
+            _ => return Err(()),
+        };
+
+        Ok(flag)
+    }
+}
+
+/// This is used by the subcommands to communicate with the option parser which flags are
+/// valid and get the result of parsing the command for flags.
+#[derive(Default)]
+struct Options<'args> {
+    null_in: bool,
+    null_out: bool,
+    quiet: bool,
+
+    invert_valid: bool,
+    invert: bool,
+
+    relative_valid: bool,
+    relative: bool,
+
+    reverse_valid: bool,
+    reverse: bool,
+
+    unique_valid: bool,
+    unique: bool,
+
+    key: Option<&'args wstr>,
+
+    types_valid: bool,
+    types: Option<TypeFlags>,
+
+    perms_valid: bool,
+    perms: Option<PermFlags>,
+
+    arg1: Option<&'args wstr>,
+}
+
+#[inline]
+fn path_out(streams: &mut io_streams_t, opts: &Options<'_>, s: impl AsRef<wstr>) {
+    let s = s.as_ref();
+    if !opts.quiet {
+        if !opts.null_out {
+            streams
+                .out
+                .append_with_separation(s, separation_type_t::explicitly, true);
+        } else {
+            let mut output = WString::with_capacity(s.len() + 1);
+            output.push_utfstr(s);
+            output.push('\0');
+            streams.out.append(output);
+        }
+    }
+}
+
+fn construct_short_opts(opts: &Options) -> WString {
+    // All commands accept -z, -Z and -q
+    let mut short_opts = WString::from(":zZq");
+    if opts.perms_valid {
+        short_opts += L!("p:");
+        short_opts += L!("rwx");
+    }
+
+    if opts.types_valid {
+        short_opts += L!("t:");
+        short_opts += L!("fld");
+    }
+
+    if opts.invert_valid {
+        short_opts.push('v');
+    }
+    if opts.relative_valid {
+        short_opts.push('R');
+    }
+    if opts.reverse_valid {
+        short_opts.push('r');
+    }
+    if opts.unique_valid {
+        short_opts.push('u');
+    }
+
+    short_opts
+}
+
+/// Note that several long flags share the same short flag. That is okay. The caller is expected
+/// to indicate that a max of one of the long flags sharing a short flag is valid.
+/// Remember: adjust the completions in share/completions/ when options change
+const LONG_OPTIONS: [woption<'static>; 10] = [
+    wopt(L!("quiet"), no_argument, 'q'),
+    wopt(L!("null-in"), no_argument, 'z'),
+    wopt(L!("null-out"), no_argument, 'Z'),
+    wopt(L!("perm"), required_argument, 'p'),
+    wopt(L!("type"), required_argument, 't'),
+    wopt(L!("invert"), no_argument, 'v'),
+    wopt(L!("relative"), no_argument, 'R'),
+    wopt(L!("reverse"), no_argument, 'r'),
+    wopt(L!("unique"), no_argument, 'u'),
+    wopt(L!("key"), required_argument, NONOPTION_CHAR_CODE),
+];
+
+fn parse_opts<'args>(
+    opts: &mut Options<'args>,
+    optind: &mut usize,
+    n_req_args: usize,
+    args: &mut [&'args wstr],
+    parser: &mut parser_t,
+    streams: &mut io_streams_t,
+) -> Option<c_int> {
+    let cmd = args[0];
+    let mut args_read = Vec::with_capacity(args.len());
```

Comment:
Hmm there may be a misunderstanding here. The way `wgetopter_t` works is that it permutes positional arguments to the end. For example:

```
path filter path1 path2 --perm read path3
```

Here the positional arguments (path1, path2, path3) are interleaved with the option-argument (--perm read). `wgetopter_t` handles this by rearranging the arguments to be:

```
path filter --perm read path1 path2 path3
```

this allows the option (`--perm`) to be handled before the positionals. After flags are handled, the remaining arguments are all positionals, because the positionals have been moved to the end.

This is why `wgetopter_t` needs a `&mut [&wstr]`: it does not modify arguments, but it may rearrange them.

(BTW I am NOT attached to this design: it was inherited from C, then C++, and now Rust. Overdue for a replacement.)

Based on that I still believe that it is wrong to copy the arguments - not for performance (agreed the overhead is negligible) but because if the arguments are rearranged, then `w.woptind` will reference the wrong argument.

Reply:
I am fully aware that wgetoptr can end up rearranging the arguments — apologies for being unclear in my communicating there, as I know this I a matter of correctness and not performance.

> (BTW I am NOT attached to this design: it was inherited from C, then C++, and now Rust. Overdue for a replacement.)

Am I interpreting this right in that you would be open to using e.g. [`clap`](https://docs.rs/clap/latest/clap/) assuming it does not cause other breaking changes? Ofc that would only be relevant after the port is done.